File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.awt.Point;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
	
	private static int n, m, minDist;
	private static ArrayList<Point> house;
	private static ArrayList<Point> chicken;
	private static int[] selected;
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		house = new ArrayList<>();
		chicken = new ArrayList<>();
		selected = new int[m];
		minDist = Integer.MAX_VALUE;
		
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			
			for (int j = 0; j < n; j++) {
				int num = Integer.parseInt(st.nextToken());
				if (num == 2) {
					chicken.add(new Point(i, j));
				} else if (num == 1) {
					house.add(new Point(i, j));					
				}
			}
		}
		
		combination(0, 0);
		
		System.out.println(minDist);
	}
	
	private static void combination(int start, int depth) {
		if (depth == m) {
			minDist = Math.min(minDist, getMinDist(selected));
			return;
		}
		
		for (int i = start; i < chicken.size(); i++) {
			selected[depth] = i;
			combination(i + 1, depth + 1);
		}
	}
	
	private static int getMinDist(int[] selected) {
		int sum = 0;
		
		for (Point h : house) {
			int min = Integer.MAX_VALUE;
			
			for (int s : selected) {
				Point c = chicken.get(s);
				min = Math.min(min, getDist(h.x, h.y, c.x, c.y));				
			}
			
			sum += min;
		}
		
		return sum;
	}
	
	private static int getDist(int x1, int y1, int x2, int y2) {
		return Math.abs(x2 - x1) + Math.abs(y2 - y1);
	}
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.kt
import kotlin.math.abs
import kotlin.math.min

private lateinit var chicken: MutableList<Pos>
private lateinit var house: MutableList<Pos>
private lateinit var chosen_chicken: MutableList<Pos>
private lateinit var chosen_chickens: MutableList<List<Pos>>
private var m: Int = 0

fun main() = java.io.StreamTokenizer(System.`in`.bufferedReader()).run {
    fun i(): Int { nextToken();return nval.toInt() }
    val n = i(); m = i(); chicken = mutableListOf(); house = mutableListOf()

    for (i in 1..n) {
        for (j in 1..n) {
            when (i()) {
                1 -> house += Pos(i, j)
                2 -> chicken += Pos(i, j)
            }
        }
    }

    chosen_chicken = mutableListOf(); chosen_chickens = mutableListOf()
    chooseChickenPos(depth = m)

    println(calcCityChickenDistance())
}

fun chooseChickenPos(cnt: Int = 0, begin: Int = 0, depth: Int) {
    if (cnt == depth) {
        chosen_chickens.add(chosen_chicken.toList())
        return
    }

    for (i in begin until chicken.size) {
        chosen_chicken.add(chicken[i])
        chooseChickenPos(cnt + 1, i + 1, depth)
        chosen_chicken.remove(chicken[i])
    }
}

fun calcCityChickenDistance(): Int {
    val chickenDistances = mutableListOf<Int>()
    var minCityChickenDistance = Int.MAX_VALUE
    var currentCityChickenDistance = 0

    chosen_chickens.forEach { cPositions ->
            house.forEach { hPos ->
                cPositions.forEach { cPos ->
                    chickenDistances += abs(hPos.x - cPos.x) + abs(hPos.y - cPos.y)
                }
                currentCityChickenDistance += chickenDistances.min()
                chickenDistances.clear()
            }
            minCityChickenDistance = min(minCityChickenDistance, currentCityChickenDistance)
            currentCityChickenDistance = 0
        }

    return minCityChickenDistance
}

data class Pos(val x: Int = 0, val y: Int = 0)